### PR TITLE
Move managing OIDC elsticsearch user configmap and secret from operator

### PR DESCRIPTION
**Pull request by Suraiya-Hameed**
## Description

Move managing of OIDC Elsticsearch users ConfigMap and Secret from operator to kube-controller. 
ConfigMap `tigera-known-oidc-users` and Secret `tigera-oidc-users-elasticsearch-credentials`, needs to updated by few other internal components and doesn't fit well into tigera-operator design.

## For PR author

- [ ] Tests for change.
- [ ] If changing pkg/apis/, run `make gen-files`
- [ ] If changing versions, run `make gen-versions`

## For PR reviewers

A note for code reviewers - all pull requests must have the following:

- [ ] Milestone set according to targeted release.
- [ ] Appropriate labels:
  - `kind/bug` if this is a bugfix.
  - `kind/enhancement` if this is a a new feature.
  - `enterprise` if this PR applies to Calico Enterprise only.

### CODE REVIEW

**Comment by asincu** [comment]
@Suraiya-Hameed Why is this change needed ?

File: pkg/render/logstorage.go
Comment (asincu) [inline pkg/render/logstorage.go:1462]: Why are we increasing access to all verbs if we are no longer watching the config map ?

File: pkg/render/logstorage.go
Comment (asincu) [inline pkg/render/logstorage.go:1468]: Why are we increasing access to all verbs ?

File: pkg/render/logstorage.go
Comment (Suraiya-Hameed) [reply pkg/render/logstorage.go:1462]: We are moving the code to manage this ConfigMap and Secret over to kube-controller, I've added updated this PRs description and added more details in the kube-controller PR.

Comment (asincu) [reply pkg/render/logstorage.go:1462]: I see we have create, delete, get & watch. Sounds good.

File: pkg/render/logstorage.go
Comment (asincu) [reply pkg/render/logstorage.go:1468]: Same as above.

**Review by asincu — APPROVED** [review]

File: pkg/controller/utils/elasticsearch.go
Comment (rene-dekker) [inline pkg/controller/utils/elasticsearch.go:221]: There are places where the es.client is used directly, like ` es.client.XPackIlmGetLifecycle().Policy(policyName).Do(ctx)`. You probably want to put access to that client in a method or it could turn into a panic.

Comment (rene-dekker) [inline pkg/controller/utils/elasticsearch.go:221]: `es.client.XPackIlmGetLifecycle().Policy(policyName).Do(ctx)` still tries to access client directly. I think you should turn it into `es.client()...` to make sure you can never use a method while es == nil.

**Comment by tmjd** [comment]
What uses the ConfigMap and Secret? I'm wondering if we need an annotation on a pod so that when they change/update the pod gets restarted to pick up those changes.
Should the operator continue to watch these resources and block logstorage if they don't exist?

**Comment by tmjd** [comment]
Do the ILM changes need to be in this PR? It seems like the changes are not related and it would be easier to review if they were separate.
If it is necessary they are in this PR could you explain why since I didn't understand.

**Comment by rene-dekker** [comment]
@tmjd That is partially my fault. I saw some inefficiencies with locking and suggested to clean that up. That led to the refactoring.

**Comment by Suraiya-Hameed** [comment]
> What uses the ConfigMap and Secret? I'm wondering if we need an annotation on a pod so that when they change/update the pod gets restarted to pick up those changes.

This ConfigMap and Secret are managed by kube-controller, it watches and re-creates it as needed, and es-proxy writes into the ConfigMap and reads from Secret. We are already passing the Elasticsearch license type to kube-controller, and the kube-controller creates/deletes this ConfigMap and Secret based on that environment value. 

> Should the operator continue to watch these resources and block logstorage if they don't exist?

That's a good idea, consequence of not having this ConfigMap and Secret is that SSO-like functionality will not work when moving from ManagerUI to Kibana. It does make sense to block logstorage till they are available.

**Comment by Suraiya-Hameed** [comment]
> Do the ILM changes need to be in this PR? It seems like the changes are not related and it would be easier to review if they were separate.
If it is necessary they are in this PR could you explain why since I didn't understand.

The changes to elasticsearch file can be a separate PR, it is mostly refactoring and has a fix to handle scenario where Elasticsearch password changes in the background.

File: pkg/controller/logstorage/logstorage_controller.go
Comment (tmjd) [inline pkg/controller/logstorage/logstorage_controller.go:544]: I'm not sure anymore if we need to block LogStorage on the OIDC resources. Is them not being present going to be a problem with LogStorage or just for the manager? Or are they used by Kibana which is launched by LogStorage.
You mentioned that going from Manager UI to Kibana wouldn't work when using SSO, but it isn't even possible to get to Kibana until Manager is available anyway.
But I guess if SSO is changed then LogStorage should be Degraded if there is a known resource missing that Kibana would need.

Assuming we should keep this check, can you add a comment here summarizing the reason why we're checking and degrading LogStorage if these resources don't exist.

File: pkg/render/logstorage.go
Comment (tmjd) [inline pkg/render/logstorage.go:1465]: We should specify the Verbs explicitly instead of using the wildcard. We haven't been very consistent about this so far but that is what we should be striving for.

File: pkg/render/logstorage.go
Comment (tmjd) [inline pkg/render/logstorage.go:1471]: Same here the verbs should be explicit.

**Review by tmjd — COMMENTED** [review]
Sorry a few more comments and requests

File: pkg/controller/logstorage/logstorage_controller.go
Comment (Suraiya-Hameed) [reply pkg/controller/logstorage/logstorage_controller.go:544]: Yes, right, if those resources are missing SSO to Kibana will be affected, and Kibana is part of LogStorage so it makes sense to degrade it.

**Review by tmjd — APPROVED** [review]

**Comment by marvin-tigera** [comment]
OK, I will merge the pull request when it's ready, leave the commits as is when I merge it, and leave the branch after I've merged it.

**Pull request merged**
